pyfstat/core.py

```diff
@@ -1649,57 +1649,46 @@ def get_semicoherent_twoF(
-        #    log10_BSGL = lalpulsar.ComputeBSGL(twoF, self.twoFX,
-        #                                       self.BSGLSetup)
-        #    return log10_BSGL/np.log10(np.exp(1))
+        twoF_per_segment = self._get_per_segment_twoF()
+        twoF = twoF_per_segment.sum()
 
-        det_stat_per_segment = self._get_per_segment_det_stat()
-        detStat = det_stat_per_segment.sum()
-
-        if np.isnan(detStat):
-            logging.debug("NaNs in semi-coherent detection statistic treated as zero")
-            det_stat_per_segment = np.nan_to_num(det_stat_per_segment, nan=0.0)
-            detStat = det_stat_per_segment.sum()
+        if np.isnan(twoF):
+            logging.debug(
+                "NaNs in per-segment 2F treated as zero"
+                " and semi-coherent 2F re-computed."
+            )
+            twoF_per_segment = np.nan_to_num(twoF_per_segment, nan=0.0)
+            twoF = twoF_per_segment.sum()
 
         if record_segments:
-            self.detStat_per_segment = det_stat_per_segment
-
-        return detStat
-
-    def _get_per_segment_det_stat(self):
-        FS = lalpulsar.ComputeTransientFstatMap(
-            self.FstatResults.multiFatoms[0], self.semicoherentWindowRange, False
-        )
+            self.detStat_per_segment = twoF_per_segment
 
         if self.BSGL is False:
-            d_detStat = 2 * FS.F_mn.data[:, 0]
-        # FIXME: BSGL should not be compute per segment, but for summed F!
+            return twoF
         else:
-            FstatResults_single = copy.copy(self.FstatResults)
-            FstatResults_single.lenth = 1
-            FstatResults_single.data = self.FstatResults.multiFatoms[0].data[0]
-            FS0 = lalpulsar.ComputeTransientFstatMap(
-                FstatResults_single.multiFatoms[0], self.semicoherentWindowRange, False,
-            )
-            FstatResults_single.data = self.FstatResults.multiFatoms[0].data[1]
-            FS1 = lalpulsar.ComputeTransientFstatMap(
-                FstatResults_single.multiFatoms[0], self.semicoherentWindowRange, False,
-            )
-
-            self.twoFX[0] = 2 * FS0.F_mn.data[0][0]
-            self.twoFX[1] = 2 * FS1.F_mn.data[0][0]
-            log10_BSGL = lalpulsar.ComputeBSGL(
-                2 * FS.F_mn.data[0][0], self.twoFX, self.BSGLSetup
-            )
-            d_detStat = log10_BSGL / np.log10(np.exp(1))
-
-        return d_detStat
+            for X in range(self.FstatResults.numDetectors):
+                FstatResults_single = copy.copy(self.FstatResults)
+                FstatResults_single.numDetectors = 1
+                FstatResults_single.data = self.FstatResults.multiFatoms[0].data[X]
+                FSX = lalpulsar.ComputeTransientFstatMap(
+                    FstatResults_single.multiFatoms[0],
+                    self.semicoherentWindowRange,
+                    False,
+                )
+                self.twoFX[X] = 2 * FSX.F_mn.data[0][0]
```

Comment:
Yes, absolutely! That should have been the whole point of this PR... :facepalm:

Reply:
Please double-check latest version, I've also added another NaN check here similar to the main one because I'm not sure how ugly the `lalpulsar.ComputeBSGL()` failure would be on NaN inputs.

(It's actually a bit dirty to just set individual segments to 0 in both cases, because then the output will be a sufficiently sane-looking semicoherent stat; while if the whole stat were nan-to-0 converted then at least an MCMC sampler would always reject the point and in a grid output the points would be more suspicious. But I'm ok with it for the moment, since there is the debug message.)

Reply:
Looks Ok to me